packages/mobile/src/exchange/saga.ts

```diff
@@ -1,5 +1,248 @@
+  Logger.debug(TAG, 'Calling @doFetchExchangeRate')
+
+  // If makerAmount and makerToken are given, use them to estimate the exchange rate,
+  // as exchange rate depends on amount sold. Else default to present large sell amount.
+  const goldMakerAmount =
+    makerAmount && makerToken === CURRENCY_ENUM.GOLD ? makerAmount : LARGE_GOLD_SELL_AMOUNT_IN_WEI
+  const dollarMakerAmount =
+    makerAmount && makerToken === CURRENCY_ENUM.DOLLAR
+      ? makerAmount
+      : LARGE_DOLLARS_SELL_AMOUNT_IN_WEI
+
+  try {
+    yield call(getConnectedAccount)
+
+    const contractKit = newKitFromWeb3(web3)
+    const exchange = yield call([contractKit.contracts, contractKit.contracts.getExchange])
+
+    const dollarMakerExchangeRate: BigNumber = yield call(
+      [exchange, exchange.getUsdExchangeRate],
+      dollarMakerAmount
+    )
+    const goldMakerExchangeRate: BigNumber = yield call(
+      [exchange, exchange.getGoldExchangeRate],
+      goldMakerAmount
+    )
+
+    if (!dollarMakerExchangeRate || !goldMakerExchangeRate) {
+      Logger.error(TAG, 'Invalid exchange rate')
+      throw new Error('Invalid exchange rate')
+    }
+
+    Logger.debug(
+      TAG,
+      `Retrieved exchange rate: 
+      ${dollarMakerExchangeRate.toString()} gold per dollar, estimated at ${dollarMakerAmount}
+      ${goldMakerExchangeRate.toString()} dollar per gold, estimated at ${goldMakerAmount}`
+    )
+
+    yield put(
+      setExchangeRate({
+        goldMaker: goldMakerExchangeRate.toString(),
+        dollarMaker: dollarMakerExchangeRate.toString(),
+      })
+    )
+  } catch (error) {
+    Logger.error(TAG, 'Error fetching exchange rate', error)
+    yield put(showError(ErrorMessages.EXCHANGE_RATE_FAILED))
+  }
+}
+
+export function* exchangeGoldAndStableTokens(action: ExchangeTokensAction) {
+  Logger.debug(`${TAG}@exchangeGoldAndStableTokens`, 'Exchanging gold and stable token')
+  const { makerToken, makerAmount } = action
+  Logger.debug(TAG, `Exchanging ${makerAmount.toString()} of token ${makerToken}`)
+  let txId: string | null = null
+  try {
+    const account: string = yield call(getConnectedUnlockedAccount)
+    const exchangeRatePair: ExchangeRatePair = yield select(
+      (state: RootState) => state.exchange.exchangeRatePair
+    )
+    const exchangeRate = getRateForMakerToken(exchangeRatePair, makerToken)
+    if (!exchangeRate) {
+      Logger.error(TAG, 'Invalid exchange rate from exchange contract')
+      return
+    }
+    if (exchangeRate.isZero()) {
+      Logger.error(TAG, 'Cannot do exchange with rate of 0. Stopping.')
+      throw new Error('Invalid exchange rate')
+    }
+
+    txId = yield createStandbyTx(makerToken, makerAmount, exchangeRate, account)
+
+    const goldTokenContract: GoldTokenType = yield call(getGoldTokenContract, web3)
+    const stableTokenContract: StableTokenType = yield call(getStableTokenContract, web3)
+    const exchangeContract: ExchangeType = yield call(getExchangeContract, web3)
```

Comment:
Can this be converted to contact kit too?

Reply:
These contracts are used for the exchange transaction. I plan to migrate these along with other state changing functions in a future PR